### PR TITLE
EICNET-1678: As a SA/SCM I must provide a reason to block a group (code refactor)

### PR DESCRIPTION
**Pull request by 4hardfire**
### Re-factored

- Create new flag "Request block (Group)" to replace with old "Block group" flag;
- Create new handler service (BlockRequestHandler) to replace with old "EntityBlockHandler";
- Re-factor NewRequestForm class so that it automatically triggers the block request accept when the request is made by an SA/SCM;
- Create email notification message to notify block request accepted and trigger notification for group owner;
- Show block operation in group header and admin lists;
- Fix coding standards.

### Tests

- [x] As SA/SCM, go to a group page of a public group and make sure you see the link "Block" in the group header
- [x] As GO/GA/GM, make sure the "Block" link is not visible and you cannot access the form to block the group -> `/request/block/group/<group-id>`
- [x] As SA/SCM, go to the form to block the group (via `/request/block/group/<group-id>` or click on "Block" link in the group header).
- [x] Provide a reason message a click on "Block group"
- [x] Make sure the group is blocked (you can go to the group edit form and check the current state)
- [x] Make sure you cannot access the form to block the group anymore
- [x] Go to the group edit form and change the status from blocked to published
- [x] Make sure you can now access the form to block the group again
- [x] As SA/SCM/GO/GA make sure you can access the page to view the Blocked history of the group -> `/groups/<group-name>/blocked-history`

### Todo:

Remove old classes (DONE):

- ~EntityBlockForm~
- ~EntityBlockRouteProvider~
- ~EntityBlockHandler~
- ~BlockFlagTypes~
- ~EntityBlockMessageCreator~
- ~Reference to notification message `NotificationMessageTemplates::ENTITY_BLOCKED`~

Remove old flags (DONE):

- ~block_group~

Remove old message templates (DONE):

- ~notify_entity_blocked~

### CODE REVIEW

File: lib/modules/eic_flags/src/Form/NewRequestForm.php [hunk 1]
```diff
@@ -74,19 +76,55 @@ public function getFormId() {
...
-      ['@action' => $action]
-    );
+    $description = NULL;
+
+    switch ($this->requestHandler->getType()) {
+      case RequestTypes::BLOCK:
+        $description = $this->t(
+          "You're about to block this @entity_type. Are you sure?",
+          [
+            '@entity_type' => $this->entity->getEntityTypeId(),
```
Comment (lramarojaona) [inline lib/modules/eic_flags/src/Form/NewRequestForm.php:86]: I guess you can grab the entity type label like this:

```php
$this->entity->getEntityType()->getLabel()
```

File: lib/modules/eic_flags/src/Form/NewRequestForm.php [hunk 2]
```diff
@@ -74,19 +76,55 @@ public function getFormId() {
...
-    $action = $this->requestHandler instanceof ArchiveRequestHandler ? 'archival' : 'deletion';
-    return $this->t('Request @action', ['@action' => $action]);
+    $description = NULL;
+
+    switch ($this->requestHandler->getType()) {
+      case RequestTypes::BLOCK:
+        $description = $this->t(
+          'Block @entity_type',
+          [
+            '@entity_type' => $this->entity->getEntityTypeId(),
```
Comment (lramarojaona) [inline lib/modules/eic_flags/src/Form/NewRequestForm.php:116]: same here:

`$this->entity->getEntityType()->getLabel()`

File: lib/modules/eic_flags/src/Form/NewRequestForm.php [hunk 3]
```diff
@@ -74,19 +76,55 @@ public function getFormId() {
...
+        $description = $this->t(
+          "You're about to block this @entity_type. Are you sure?",
+          [
+            '@entity_type' => $this->entity->getEntityTypeId(),
+          ],
+        );
+        break;
+
+      default:
+        $action = $this->requestHandler instanceof ArchiveRequestHandler ? 'archival' : 'deletion';
```
Comment (lramarojaona) [inline lib/modules/eic_flags/src/Form/NewRequestForm.php:92]: Should we have a helper function to retrive this?

As this code is duplicated on lines 122 and 142 as well

File: lib/modules/eic_flags/src/Form/NewRequestForm.php [hunk 4]
```diff
@@ -95,12 +133,24 @@ public function getConfirmText() {
...
+    }
+
     $form['reason'] = [
       '#type' => 'textarea',
       '#title' => $this->t(
-        'Please explain why this content should be @action',
-        ['@action' => $action]
+        'Please explain why this @entity_type should be @action',
+        [
+          '@entity_type' => $this->entity->getEntityTypeId(),
```
Comment (lramarojaona) [inline lib/modules/eic_flags/src/Form/NewRequestForm.php:151]: `$this->entity->getEntityType()->getLabel()`

File: lib/modules/eic_flags/src/Form/NewRequestForm.php [hunk 5]
```diff
@@ -128,6 +178,29 @@ public function validateForm(array &$form, FormStateInterface $form_state) {
...
+    $entity_moderation_state = $this->entity->get('moderation_state')->value;
+
+    // Group is not published, so we do nothing.
+    if ($entity_moderation_state === BlockRequestHandler::ENTITY_BLOCKED_STATE) {
+      $form_state->setError(
+        $form,
+        $this->t(
+          'This @entity_type is already blocked.',
+          [
+            '@entity_type' => $this->entity->getEntityTypeId(),
```
Comment (lramarojaona) [inline lib/modules/eic_flags/src/Form/NewRequestForm.php:199]: `$this->entity->getEntityType()->getLabel()`

File: lib/modules/eic_flags/src/Form/NewRequestForm.php [hunk 6]
```diff
@@ -140,7 +213,22 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
...
+
+    // If the request was to block a group, we show a different message status.
+    if (
+      $this->entity->getEntityTypeId() === 'group' &&
+      $this->requestHandler->getType() === RequestTypes::BLOCK
+    ) {
+      $status_message = $this->t(
+        'The @entity_type has been blocked.',
+        [
+          '@entity_type' => $this->entity->getEntityTypeId(),
```
Comment (lramarojaona) [inline lib/modules/eic_flags/src/Form/NewRequestForm.php:226]: $this->entity->getEntityType()->getLabel()

File: lib/modules/eic_messages/src/Service/RequestMessageCreator.php [hunk 7]
```diff
@@ -93,7 +94,7 @@ public function requestInsert(
   ) {
     $handler = $this->collector->getHandlerByType($type);
     if (!$handler instanceof HandlerInterface) {
-      Drupal::logger('eic_messages')->warning(
+      \Drupal::logger('eic_messages')->warning(
```
Comment (lramarojaona) [inline lib/modules/eic_messages/src/Service/RequestMessageCreator.php:97]: I know it's not really part of this ticket, but here we could use `LoggerChannelTrait`

See https://www.drupal.org/docs/8/api/logging-api/overview

File: lib/modules/eic_flags/src/Form/NewRequestForm.php (hunk 6 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Form/NewRequestForm.php]: Done!

File: lib/modules/eic_flags/src/Form/NewRequestForm.php (hunk 5 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Form/NewRequestForm.php]: Done!

File: lib/modules/eic_flags/src/Form/NewRequestForm.php (hunk 2 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Form/NewRequestForm.php]: Done!

File: lib/modules/eic_flags/src/Form/NewRequestForm.php (hunk 1 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Form/NewRequestForm.php]: Done!

File: lib/modules/eic_flags/src/Form/NewRequestForm.php (hunk 3 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Form/NewRequestForm.php:92]: @lramarojaona To be honest, this part of code was already like this in the past. I don't think it's being used anywhere else and I don't see a big issue leaving it as it is for now.
Small remark: in fact, line 122 is duplicated of 92 and not 142. Line 142 shows the action text in the "past" (deleted/archived and not archival/deletion).
If you think it's really useful I can create a public static function then.

Comment (lramarojaona) [reply lib/modules/eic_flags/src/Form/NewRequestForm.php:92]: Ok you're right, maybe add a @todo for later?

Comment (4hardfire) [reply lib/modules/eic_flags/src/Form/NewRequestForm.php:94]: Done!

**Review by lramarojaona — APPROVED** [review]

**Pull request merged**